### DISCUSSION TITLE
Gaussian White Noise Generation

### DISCUSSION THREAD

**Discussion by franjf97** [pschatzmann/arduino-audio-tools#427] [Q&A] [answered]
Hi everyone! I've been able to use the GeneratedSoundStream class, with NoiseGenerator to produce noise as the ouput sound in a I2S DAC that I'm using as an audio source for some experiments. However, I've seen that the generation of this stream is made using the "rand()" function, which produces an uniform probability distribution as I'm aware. This isn't a problem in many cases, but for the application I'm researching I'd like to have white noise with gaussian distribution.

Doing some research, I´ve found that there's an algorithm called "Box-Muller Transform", that converts random samples with uniform distribution to normal distribution using a simple equation. Here is an article that explains all of this:

https://www.baeldung.com/cs/uniform-to-normal-distribution

I tried to modify the NoiseGenerator class to implement this algorithm but without success. It generated some output but the audio seemed to be broken, maybe because the operations involved (log, sqrt) are too slow, but this is just an assumption...

This was my test code, located in src/AudioEffects/SoundGenerator.h:

```
template <class T>
class NoiseGenerator : public SoundGenerator<T> {
    public:
        /// the scale defines the max value which is generated
        NoiseGenerator(float amplitude = 32767.0) {
            this->amplitude = amplitude;
        }

        /// Provides a single sample
        T readSample() {
            float u1, u2;
            float z;
            T result;
            u1=(float)rand() / (float)RAND_MAX ;
            u2=(float)rand() / (float)RAND_MAX ;
            z=sqrt(-2*log(u1))*cos(2*PI*u2);      
            result=amplitude*z;      
            return result;
        }

    protected:
        float amplitude;
        // //range : [min, max]
        int random(int min, int max)  {
            return min + rand() % (( max + 1 ) - min);
        }
};   
``` 

Does anyone come up with an idea? It would be very helpful.

Thanks!

**Comment by pschatzmann** [comment, marked as answer]
I suggest to plot the result to a CsvStream so that you can visually check if the data looks random or if there is another issue. All data should be between -32767.0 and +32767.0
If your assumption is right you can try to lower the sample rate or you can measure the number of samples that you can generate in 1 second, to determine the max sample rate that you can use.

Did you already try to use the [standard c++ function](https://stackoverflow.com/questions/32889309/adding-gaussian-noise)  instead?

Finally I suggest that you create a separate class with a different name. If you have it working I would be interested in a pull request.
